algosdk/atomic_transaction_composer.py

```diff
@@ -165,67 +188,156 @@ def add_method_call(
-        additional_types = []
+        raw_values = []
+        raw_types = []
         txn_list = []
         # First app arg must be the selector of the method
         app_args.append(method.get_selector())
         # Iterate through the method arguments and either pack a transaction
         # or encode a ABI value.
         for i, arg in enumerate(method.args):
-            if arg.type in abi.method.TRANSACTION_ARGS:
+            if abi.is_abi_transaction_type(arg.type):
                 if not isinstance(method_args[i], TransactionWithSigner):
                     raise error.AtomicTransactionComposerError(
                         "expected TransactionWithSigner as method argument, but received: {}".format(
                             method_args[i]
                         )
                     )
-                if method_args[i].txn.group and method_args[i].txn.group != 0:
-                    raise error.AtomicTransactionComposerError(
-                        "cannot add a transaction with nonzero group ID"
-                    )
                 txn_list.append(method_args[i])
-            elif len(app_args) > self.MAX_ABI_APP_ARG_LIMIT:
-                # Pack the remaining values as a tuple
-                additional_types.append(arg.type)
-                additional_args.append(method_args[i])
             else:
-                encoded_arg = arg.type.encode(method_args[i])
-                app_args.append(encoded_arg)
+                if abi.is_abi_reference_type(arg.type):
+                    current_type = abi.UintType(8)
+                    if arg.type == abi.ABIReferenceType.ACCOUNT:
+                        account_arg = str(method_args[i])
+                        if account_arg == sender:
+                            current_arg = 0
+                        elif account_arg in accounts:
+                            current_arg = accounts.index(account_arg) + 1
+                        else:
+                            current_arg = len(accounts) + 1
+                            accounts.append(account_arg)
+                    elif arg.type == abi.ABIReferenceType.ASSET:
+                        asset_arg = int(method_args[i])
+                        if asset_arg in foreign_assets:
+                            current_arg = foreign_assets.index(asset_arg)
+                        else:
+                            current_arg = len(foreign_assets)
+                            foreign_assets.append(asset_arg)
+                    elif arg.type == abi.ABIReferenceType.APPLICATION:
+                        app_arg = int(method_args[i])
+                        if app_arg == app_id:
+                            current_arg = 0
+                        elif app_arg in foreign_apps:
+                            current_arg = foreign_apps.index(app_arg) + 1
+                        else:
+                            current_arg = len(foreign_apps) + 1
+                            foreign_apps.append(app_arg)
```

Comment:
nit: would be nice to consider `populateForeignArray`-like method, since the code is somewhat redundant (but it is optional)